### ISSUE TITLE
More improvements on working with optional fields

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#1079]
#926 made it more convenient to work with optional fields by making `T x?;` behave more like `T? x;` when `T` does not allow `()`.

We can take #926 a bit further.

Optional fields are still clunky with query. When a record has nillable types like this:

```
record Tnil {
   int? x;
   int? y:
};
```

then we can use query to transform a `Tnil [] v` like this

```
from var { x, y } in v where foo(y) select { x, y }
```

But this cannot be used if the type uses optional fields:

```
record Topt {
   int x?;
   int y?;
};
```

There's both a destructuring problem and constructing problem. If t has type Topt, then (following #926) this should work:

```
var { x, y } = t;
```

and should result in x and y having type `int?` i.e. it should be the same as

```
var x = t.x;
var y = t.y;
```

Similarly, on construction I should be able to do:

```
int? x = 1;
int? y = ();
Topt t = {x, y};
```

and the result should be that `t` has a `x` field and no `y` field.  If we do both of these things then this same code

```
from var { x, y } in v where foo(y) select { x, y }
```

will work when v has type `Topt`.

Similarly if r has a type with an optional field `int x?`, then

```
r.x = ();
```

should remove the field.

In other words, when we have a field  `T x?;` where `T` does not allow nil, we can consistently use nil to represent the absence of the `x` field. #926 is already doing this for one case: this extends it to others.

**Comment by hasithaa** [comment]
@jclark https://github.com/ballerina-platform/ballerina-spec/issues/928 is not the correct issue.

**Comment by jclark** [comment]
@hasithaa Whoops, should have been #926

**Comment by KavinduZoysa** [comment]
@jclark, just to confirm, if we use `member-access-lvexpr` () instead of `field-access-lvexpr` should be the behavior same as explained above?

**Comment by jclark** [comment]
@KavinduZoysa If we do that, then we should also make

```
string x = "x";
r[x] = ();
```

work.  This is starting to feel a bit of a slippery slope...

@hasithaa Any thoughts?

**Comment by hasithaa** [comment]
Since we support this only for optional fields., I think we can safely allow this for `field-access-lvexpr` and possibly allow  `member-access-lvexpr` when the expression is a constant. But the support for `member-access-lvexpr`  can be added later.

**Comment by jclark** [comment]
I am not sure about member-access-lvexpr, so I would prefer not to support this for now.

**Comment by KavinduZoysa** [comment]
@jclark, please consider the below example.

```ballerina
type Student record {|
   int id?;
|};

type School record {|
   Student student?;
|};

public function main() {
   School school = {student: {id: 2}};
   school.student.id = ();
}
```

With this change, static type of `school.student.id` becomes `int?`.  In the expression `school.student.id`, does the type of `school.student` become `Student?` ? 
If so this will be a breaking change.

**Comment by jclark** [comment]
In the above `school.student.id` is not an expression, it is an lvexpr (l-value expression) and typing works differently for these https://ballerina.io/spec/lang/master/#section_7.15.1

The result of evaluating an lvexpr refers to a cell that either contains either undefined or a value belonging to a type. For example, `school.student` refers to a cell that the compiler determines can contain undefined or value of type Student (not not nil). If you assign nil to this, it sets the cell to undefined.

Handling `school.student.id` will use a filling-read.

**Comment by KavinduZoysa** [comment]
The implementation is done only for the requirements mentioned in https://github.com/ballerina-platform/ballerina-spec/issues/1079#issue-1205250171. It will be available in the update 3 release.

**Comment by jclark** [comment]
@KavinduZoysa Please review the changes I made. If you see any problems, please create a new issue.
